File: Range Product Queries of Powers/Range Product Queries of Powers.cpp
// Solution for Range Product Queries of Powers in CPP

class Solution {
public:
    vector<int> productQueries(int n, vector<vector<int>>& queries) {
        const int MOD = 1e9 + 7; // Modulo value for large numbers

        vector<int> powersOfTwo; // Stores the powers of 2 that sum to n
        vector<int> queryResults; // Final answers for all queries

        // Step 1: Find the powers of two that sum to n
        // Each set bit in n's binary representation corresponds to a power of 2
        for (int bitPosition = 0; bitPosition < 32; bitPosition++) {
            if ((n & (1 << bitPosition)) != 0) { // Check if this bit is set
                powersOfTwo.push_back(1 << bitPosition); // Store 2^bitPosition
            }
        }

        // Step 2: Answer each query
        for (const auto& query : queries) {
            int leftIndex = query[0];   // Start index in powersOfTwo
            int rightIndex = query[1];  // End index in powersOfTwo
            long long product = 1;      // Store the product result

            // Multiply all values from leftIndex to rightIndex
            for (int idx = leftIndex; idx <= rightIndex; idx++) {
                product = (product * powersOfTwo[idx]) % MOD;
            }

            queryResults.push_back(product); // Save this query's result
        }

        return queryResults;
    }
};

File: Range Product Queries of Powers/Range Product Queries of Powers.java
// Solution for Range Product Queries of Powers in JAVA

class Solution {
    public int[] productQueries(int n, int[][] queries) {
        final int MOD = 100_00_00_007;

        List<Integer> powersOfTwo = new ArrayList<>();

        for (int bitPosition = 0; bitPosition < 32; bitPosition++) {
            if ((n & (1 << bitPosition)) != 0) {
                powersOfTwo.add(1 << bitPosition); 
            }
        }

        int q = queries.length;
        int[] queryResults = new int[q];

        for (int i = 0; i < q; i++) {
            int leftIndex = queries[i][0];  
            int rightIndex = queries[i][1]; 
            long product = 1;     

            for (int idx = leftIndex; idx <= rightIndex; idx++) {
                product = (product * powersOfTwo.get(idx)) % MOD;
            }

            queryResults[i] = (int) product; 
        }

        return queryResults;
    }
}
